add-custom-dynamic-entity-data-source/src/main/java/com/esri/arcgismaps/sample/addcustomdynamicentitydatasource/components/CustomEntityFeedProvider.kt

```diff
@@ -0,0 +1,172 @@
+package com.esri.arcgismaps.sample.addcustomdynamicentitydatasource.components
+
+import com.arcgismaps.data.Field
+import com.arcgismaps.data.FieldType
+import com.arcgismaps.geometry.Point
+import com.arcgismaps.geometry.SpatialReference
+import com.arcgismaps.realtime.CustomDynamicEntityDataSource
+import com.arcgismaps.realtime.DynamicEntityDataSourceInfo
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.NonCancellable
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.contentOrNull
+import kotlinx.serialization.json.double
+import kotlinx.serialization.json.jsonObject
+import kotlinx.serialization.json.jsonPrimitive
+import java.io.File
+import java.io.IOException
+
+class CustomEntityFeedProvider(
+    fileName: String,
+    private val entityIdField: String,
+    private val delayDuration: Long
+) :
+    CustomDynamicEntityDataSource.EntityFeedProvider {
+
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    private val observationsFile = File(fileName)
+
+    // Create a shared flow to emit feed events.
+    private val _feed = MutableSharedFlow<CustomDynamicEntityDataSource.FeedEvent>(
+        extraBufferCapacity = Int.MAX_VALUE,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+
+    // Expose the feed as a shared flow.
+    override val feed = _feed
+
+    // Keep track of the feed job to allow us to properly cancel it when needed.
+    private var feedJob: Job? = null
+
+    /**
+     * Called when the data source is connected. Checks for presence of the observations file and
+     * starts reading the file asynchronously.
+     */
+    override suspend fun onConnect() {
+        if (!observationsFile.exists()) {
+            throw IOException("Observations file does not exist.")
```

Comment:
Would it make sense to also show the connection status of the CustomDynamicEntityDataSource in the UI? If the feed fails to connect for some reason, it would be obvious why. Alternatively, just log `CustomDynamicEntityDataSource.connectionStatus` to LogCat.